### DISCUSSION TITLE
Erweiterung des Victron Shunt bezogenes Codes

### DISCUSSION THREAD

**Discussion by swingstate** [hoylabs/OpenDTU-OnBattery#539] [General]
Ich habe mich weiter in GitHub und VScode eingelesen und ein [Fork](https://github.com/helgeerbe/OpenDTU-OnBattery/compare/development...swingstate:OnBattery-dev:development) erstellt um meine Änderungen lokal zu testen. 

**Kontext**: Die Victron Smartshunt Unterstützung in Zusammenhang mit einer 2x 25,6V Batteriebank.

Mein Ansinnen ist es den Code der den Victron Shunt betrifft dahingehend zu erweitern, dass folgende Dinge aus Nutzersicht abgedeckt sind: 

- Anzeige der MIttelpunktspannung (VM) und Mittelpunktspannungsabweichung (DM) sowie der durch den Shunt gemessenen Leistung (W) im Liveview 
- Übermittlung von VM, DM, W an einen MQTT Broker

Ich würde das Batterie Config Menü des SmartShunts gerne ähnlich wie für das JK BMS anpassen und eine Checkbox einfügen, mit der der Benutzer bestätigen kann, dass die Batterie eine Bank aus mindestens 2 Batterien ist. Nur wenn diese Checkbox aktiv ist sollen dann VM und DM im Liveview angezeigt werden. 

Ich habe mich dazu am Menü für das JKBMS orientiert, jedoch wird die gewünschte Checkbox nicht angezeigt. Dies ist der Code aus der BatteryAdminView.vue:

```
            <CardElement v-show="batteryConfigList.enabled && batteryConfigList.provider == 2"
                         :text="$t('batteryadmin.ShuntBatteryMode')" textVariant="text-bg-primary" addSpace>

                <InputElement :label="$t('batteryadmin.ShuntBatterySerialConfiguration')"
                              v-model="batteryConfigList.shunt_serial_configuration"
                              type="checkbox" />

            </CardElement>
```

Ich habe dann noch weitere Anpassungen vorgenommen, die ich hier nicht alle auflisten möchte - sieht man ja gut im Fork. 

**Meine 1. Frage** ist nun - Gibt es irgendwo im Code noch weitere Punkte die angepasst werden müssen, damit sich das Menü für den Shunt entsprechend anpasst? 

**Meine 2. Frage** ist, Wo muss der Code geändert werden um im LiveView die gewünschten Text und nicht den Namen einer Variablen anzuzeigen? 

Aktuell sehe ich nur das Konstrukt aus battery + Variable, obwohl ich die `de.json` angepasst habe.  

Aus der de.json: 
```
        "midpoint": "Mittelpunktspannung",
        "midpointDIV": "Abweichung",
        "instantaneousPower": "Leistung"
```

![Bildschirmfoto 2023-11-26 um 21 41 52](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/73672554/e06bd810-494a-43b8-a2ae-0271381fdc48)

Vielleicht habt ihr ja Tipps?

**Comment by schlimmchen** [comment]
Das sieht mir stark danach aus als hättest du nicht `npm run build` (hast du `npm`, also [Node.js](https://nodejs.org), schon installiert?) im `webapp` Unterordner ausgeführt, bevor du die Firmware gebaut hast. Schau dir bei Anpassungen an der WebApp am besten `npm run dev` an und beachte den Kommentar `// example 'vite.user.ts': export const proxy_target = '192.168.16.107'` in der Datei `webapp/vite.config.ts` (damit steuerst du, wessen web sockets die Daten für die an `http://localhost:5173/` servierte WebApp liefern).

**Reply by swingstate** [reply]
Node.js habe ich nicht in Betracht gezogen sondern nur via Vscode kompiliert und geflasht. 

Danke!

**Reply by swingstate** [reply]
Bzgl. der Kompilierlogik - Bereitet `npm run build` die Views für die Firmware auf Basis der *view.vue Datei vor? Änderungen in den Views müssen also immer über npm "vorkompiliert" werden?

**Reply by schlimmchen** [reply]
Aus den Quelldateien im `webapp` Unterordner wird vpn `npm` eine Web Applikation gebaut. Diese wird in die OpenDTU Firmware einkompiliert und vom Webserver an den Webbrowser ausgeliefert. Also ja, da wird etwas "vorkompiliert" und dieses Artefakt wird dann in die Firmware eingebacken.

**Comment by 88markus88** [comment]
@swingstate Super, das wird ein richtig nützliches Feature! Das sollte aber auch 2 * 12 V Batteriebänke berücksichtigen, auch da ist die Mittelpunktspannung ein nützlicher Wert.

Wenn du dir diese Mühe machst, könntest du die Anzeigen für die Batteriewerte an die tatsächlich vom Smartshunt gelieferten Werte anpassen? Derzeit wird da unnötig abgeschnitten bzw. falsch gerundet:
Batterie % liefert der Smartshunt in Promille, Anzeige sollte dann auf 0.1% genau sein. Dazu kommt, dass derzeit 499 Promille in 49% übersetzt werden...
Analog geladene und entladene KWh: da steht in der Nachkommastelle immer eine 0, 
Ich vermute, da werden Typecasts etwas zu einfach gemacht.

**Reply by swingstate** [reply]
@schlimmchen ich arbeite momentan primär an dem Ve.Direct MQTT Thema und habe ein paar Fragen dazu.

**Erstes Thema - MQTT Details:**

Wo werden die Details der Nachkommastellen für z.B. discharged & chargedenergy abgeschnitten? Wenn ich mir die Rohdaten von H17 über MQTT senden lasse und ansehe bekomme ich z.B: 37604 angezeigt (lt. Datenblatt wird in 0.01 kWh gemessen). Der Code in den Releases in `Batterystats.cpp` macht daraus mittels `/ 100` 376 kWh.

```
void VictronSmartShuntStats::updateFrom(VeDirectShuntController::veShuntStruct const& shuntData) {
...
...
...
 _dischargedEnergy = shuntData.H17 / 100;
```

Durch 100 dividieren macht also Sinn. Nur wo bleiben dann die "04" Nachkomma Details? Via MQTT wird ja nur ein String gesendet. Dieser String muss ja vorher schon inkl. Nachkomme Details vorhanden sein, oder?

Für `getLiveViewData` kann ich zwar die Nachkommastellen auf 2 erhöhen, Beispiel: `addLiveViewValue(root, "dischargedEnergy", _dischargedEnergy, "KWh", 2);` jedoch bleibt es dann im LiveView bei zwei Nullen hinter dem Komma, obwohl dort 04 angezeigt werden müsste. 

**Zweites Thema - MQTT Autodiscovery:**

Ich würde den Abschnitt `sensor/dtu_` in `MqttHandleHass.cpp` gerne um einige Battery Felder (Midpoint, Midpoint Abweichung, Power etc.) erweitern, welche ich bereits erfolgreich über MQTT senden kann. Wo genau muss ich das machen? 

Ich habe den Eindruck dass Ve.Direct auch `MqttHandlePylontechHass.cpp` verwendet, denn nur dort finde ich den Teil des Strings der sich auch in den Ve.Direct Battery MQTT Topics zeigt `dtu_battery_`.

<img width="396" alt="Bildschirmfoto 2023-12-18 um 23 23 48" src="https://github.com/helgeerbe/OpenDTU-OnBattery/assets/73672554/bbd7a625-f0ba-4e0e-bc68-85dc850679ae">

Für Tipps wäre ich dankbar.

**Reply by schlimmchen** [reply]
@swingstate Sorry, ich habe noch nicht auf deine Fragen reagiert... War eine stressige Zeit zuletzt.

Ich sehe in deinem Fork, dass du deine Wünsche erfolgreich implementiert hast. Damit sind deine Fragen also obsolet.

Wenn du deine sinnvollen Verbesserungen in Helges Projekt einbringen willst, dann erstelle doch einen Pull Request. Allerdings ist dann definitiv Nacharbeit erforderlich, die dann im Rahmen des PR besprochen werden könnte.

**Comment by 88markus88** [comment]
Hallo ,
Die Nachkommastellen werden mit ziemlicher Sicherheit durch eine Integer
Division abgeschnitten. Wenn in

```
    shuntData.H17 / 100;
```
beide Variablen Integer sind, wird das Ergebnis auch ein Integer mit
abgeschnittenen Nachkommastellen:


~~~
    10234 / 100 ergibt dann 100
~~~
Probier doch mal mit Typecast
```
_dischargedEnergy = (float)shuntData.H17 / 100.0;
```
ggf. alternativ:
```
_dischargedEnergy =  (float)shuntData.H17 / (float)100;
```

Falls _dischargedEnergy eine floating point Variable ist, kommt dann beim
Beispiel 100.24 raus.

Grüsse, Markus


swingstate ***@***.***> schrieb am Mo., 18. Dez. 2023, 23:24:

**Reply by swingstate** [reply]
Danke Markus!

Das sieht jetzt schon besser aus. Habe für chargedEnergy eine Anpassung vorgenommen und bekomme nun zwei Nachkommastellen mit Details. 

![Bildschirmfoto 2023-12-19 um 13 16 55](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/73672554/2e248db0-65d7-451a-b021-6819b2bf61ac)

Werde das für weitere Felder anpassen und dann in meinem Fork einen Release zum testen hochladen. 
Habe diesen Stand zwei Boards (Dev & Prod) laufen.

**Comment by swingstate** [comment]
All current MQTT related changes are included in this test release: 

https://github.com/swingstate/OnBattery-dev/releases/tag/2023.12.19-beta

![Bildschirmfoto 2023-12-19 um 16 52 36](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/73672554/a4058e39-9660-4dcc-86ae-dc3d239bd348)

**Reply by swingstate** [reply]
After some trial and error I was able to add the data to the MQTT autodiscovery.

Details can be found in the fork [here](https://github.com/swingstate/OnBattery-dev/releases/tag/beta), including a .bin file to update existing installations: 

Screenshots from Homeassistant: 
![Bildschirmfoto 2023-12-20 um 19 43 58](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/73672554/8990fc11-e377-41c7-93a6-0e388e8e23c4)

![Bildschirmfoto 2023-12-20 um 19 44 57](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/73672554/55a7686e-3b41-4381-a491-0d11d072341a)
